runtime/tcp.cpp

```diff
@@ -0,0 +1,333 @@
+  }
+  if (!fd.has_value()) {
+    php_warning("stream_socket_client() : Unable to connect to %s (%s)", url.c_str(), error_description.c_str());
+    return false;
+  }
+
+  Stream stream = url;
+  stream.append(string("-fd-")).append(f$strval(fd.val()));
+  details::register_socket(stream, fd.val());
+  return stream;
+}
+
+bool tcp_check_correct(const Stream &stream) {
+  if (dl::query_num != opened_tcp_client_sockets_last_query_nam || !opened_tcp_client_sockets->has_key(stream.to_string())) {
+    php_warning("TCP socket \"%s\" is not opened", stream.to_string().c_str());
+    return false;
+  }
+
+  return true;
+}
+
+int tcp_get_fd(const Stream &stream) {
+  return tcp_check_correct(stream) ? opened_tcp_client_sockets->get_value(stream.to_string()).fd : -1;
+}
+
+Optional<int64_t> tcp_fwrite(const Stream &stream, const string &data) {
+  if (!tcp_check_correct(stream)) {
+    return {};
+  }
+
+  FILE *src = opened_tcp_client_sockets->get_value(stream).src;
+  dl::enter_critical_section();
+  int res = fwrite(data.c_str(), sizeof(char), data.size(), src);
+  dl::leave_critical_section();
+
+  if (res == -1) {
+    if (errno != EAGAIN) {
+      php_warning("fwrite() : An error occurred while sending TCP-package");
+      return {};
+    } else {
+      return 0;
+    }
+  }
+  return res;
+}
+
+Optional<string> tcp_fread(const Stream &stream, int64_t length) {
+  if (!tcp_check_correct(stream)) {
+    return {};
+  }
+
+  string data(length, false);
+  FILE *src = opened_tcp_client_sockets->get_value(stream).src;
+  dl::enter_critical_section();
+  int res = fread(data.buffer(), sizeof(char), length, src);
+  dl::leave_critical_section();
+
+  if (res == -1) {
+    if (errno != EAGAIN) {
+      php_warning("fread() : An error occurred while receiving TCP-package");
+      return {};
+    } else {
+      return "";
+    }
+  }
+  return data.substr(0, res);
+}
+
+Optional<string> tcp_fgets(const Stream &stream, int64_t length) {
+  if (!tcp_check_correct(stream)) {
+    return {};
+  }
+  if (length < 0) {
+    length = getpagesize();
+  }
+
+  string data(length, false);
+  FILE *src = opened_tcp_client_sockets->get_value(stream).src;
+  dl::enter_critical_section();
+  char *res = fgets(data.buffer(), length, src);
+  dl::leave_critical_section();
+
+  if (res == nullptr) {
+    return {};
+  }
+  string::size_type end_pos = data.find(string("\n")) + 1;
+  return data.substr(0, end_pos);
+}
+
+Optional<string> tcp_fgetc(const Stream &stream) {
+  if (!tcp_check_correct(stream)) {
+    return {};
+  }
+
+  FILE *src = opened_tcp_client_sockets->get_value(stream).src;
+  dl::enter_critical_section();
+  int res = fgetc(src);
+  dl::leave_critical_section();
+
+  if (res == -1) {
+    return {};
+  }
+  return f$chr(res);
+}
+
+bool tcp_fclose(const Stream &stream) {
+  int fd = tcp_get_fd(stream);
+
+  dl::enter_critical_section();
+  int res = close(fd);
```

Comment:
Why not `fclose` o be consistent with other places?

Reply:
fix